actionview/lib/action_view/renderer/collection_renderer.rb

```diff
@@ -0,0 +1,188 @@
+# frozen_string_literal: true
+
+require "action_view/renderer/partial_renderer"
+
+module ActionView
+  class PartialIteration
+    # The number of iterations that will be done by the partial.
+    attr_reader :size
+
+    # The current iteration of the partial.
+    attr_reader :index
+
+    def initialize(size)
+      @size  = size
+      @index = 0
+    end
+
+    # Check if this is the first iteration of the partial.
+    def first?
+      index == 0
+    end
+
+    # Check if this is the last iteration of the partial.
+    def last?
+      index == size - 1
+    end
+
+    def iterate! # :nodoc:
+      @index += 1
+    end
+  end
+
+  class CollectionRenderer < PartialRenderer # :nodoc:
+    include ObjectRendering
+
+    class CollectionIterator # :nodoc:
+      include Enumerable
+
+      def initialize(collection)
+        @collection = collection
+      end
+
+      def each(&blk)
+        @collection.each(&blk)
+      end
+
+      def size
+        @collection.size
+      end
+    end
+
+    class SameCollectionIterator < CollectionIterator # :nodoc:
+      def initialize(collection, path, variables)
+        super(collection)
+        @path      = path
+        @variables = variables
+      end
+
+      def from_collection(collection)
+        self.class.new(collection, @path, @variables)
+      end
+
+      def each_with_info
+        return enum_for(:each_with_info) unless block_given?
+        variables = [@path] + @variables
+        @collection.each { |o| yield(o, variables) }
+      end
+    end
+
+    class PreloadCollectionIterator < SameCollectionIterator # :nodoc:
```

Comment:
Do you think it is better to have Active Record specific behavior in Action View? I can clearly see how this improved the code but at some time this is the first time we have Active Record specific behavior inside Action View. It will also call the conditional every time a collection is passed even if the collection is the an Active Record collection.

One alternative is extract the conditional:

```
      collection = if collection.respond_to?(:preload_associations)
        PreloadCollectionIterator.new(collection, partial, iter_vars, collection)
      else
        SameCollectionIterator.new(collection, partial, iter_vars)
      end
```

to a method and only override that method if Active Record is present.

I'm not strong about this either so I'll go with what you decided. I like the way you implemented this as well.

Reply:
> Do you think it is better to have Active Record specific behavior in Action View? I can clearly see how this improved the code but at some time this is the first time we have Active Record specific behavior inside Action View.

I thought about that, but we're not referencing any constants from Active Record, only checking to see if the collection responds to `preload_associations`. Any collection could technically support preloading (though I guess  only Active Record does 😬).

My thought process was basically:

* ActionView is usually used with ActiveRecord, so most people won't pay a penalty (this is also why I included the module [in the benchmark](https://gist.github.com/tenderlove/8be9006e1d410398afdad30f4657e666#file-render_benchmark-rb-L10)
* Collection rendering is probably not a bottleneck in apps that use AV but don't use AR, so the extra conditional probably doesn't matter
* Maybe other collection types would like the opportunity to implement preloading, and this decouples us from ActiveRecord specifically

Anyway, I don't have a strong preference, but I do like that we could remove that module, the initializer, and decouple from AR (kind of).

Reply:
👍